[QA Prep Required][Migration Required] Feature/Guardian on Nodes + Groups Backend [PLAT-1041][PLAT-1185][PLAT-1268]

pattisdr opened the pull request:
❗️ This PR is almost half tests, and a large percentage of the files changed: turning a string into a constant, or changes to expected input in the add_contributor method.  Actual additions closer to 3,000 lines, with new Groups endpoints, groups models, query rewrites, and guardian changes.

Drawings for permission organization after this PR:
[Node Guardian Structure](https://docs.google.com/drawings/d/1gEM6KMYKgfkWfP0OauefGk9yIOXl-m7DMfwVumcbuV0/edit)
[OSF Groups Structure](https://docs.google.com/drawings/d/1ZlAiz9h6eqI6Peu68xGFBhkHLbAHmuD7qyBDIJhscZU/edit?usp=sharing
)

[Next steps](https://openscience.atlassian.net/browse/PLAT-1264) Some changes that are pending Groups Front end being added
## Purpose
1.  Changes what tables we use to track permissions for Nodes - moved to django-guardian.
2.  Adds the backend for OSF Groups - allows you to add a collection of users to multiple projects - numerous API endpoints added to expose this functionality.
 
## Changes
### Part I Django-Guardian
- We are now using django-guardian for AbstractNodes - guardian lets us add per-object permissions.  Previously we were tracking `read`, `write`, and `admin` columns in the Contributor table, now we're using direct foreign key guardian tables for AbstractNodes, Preprints, and OSFGroups. Direct foreign keys speed up our queries.
- How guardian works on Nodes: Basically, every time an AbstractNode is created, a `read`, `write` and `admin` Django group is created for it. Each Django group is given permissions to that node specifically (this is what the guardian tables are for; they allow us to give a Django group a specific permission to a specific node). The `admin` group is given admin, write, and read permissions to the node, the `write` group is just given write/read perms, and the `read` group just has read perms.  To give a user, then, write permissions to a node, you would add that user to the Node's `write` Django group.  If the user belongs to the write group, they inherit read/write permissions.   Giving permissions to the Django group and then adding the users to these groups (rather than adding perms directly to users) makes it really easy to tell which users have which perms, as well as making it easy to quickly remove multiple user's perms.
- As a result of adding guardian, many of our Django queries and SQL queries had to be rewritten. Whereas before, we might check to see that the write contributor column was True, to see if the user can edit, now we check to see if the user has write permissions. This works well with the addition of the OSFGroups feature, as this will return True if they have write permissions because they're a contributor, or write permissions because they're a group member. 
### Part II OSF Groups
- Adds an OSFGroup model (which also uses guardian for its own permissions). OSFGroups are collections of users, like my science lab, and I want to add my lab to many projects at once. 
- How guardian works on OSFGroups: When an OSFGroup is created, a Django member and Django manager group are created for it.  The member group is given member permissions, and the manager group is given manager permissions. The OSFGroup's creator is added to both the manager and member group.  Managers belong to both groups, members belong to the member group only. 
- You must be a group manager to manage group membership. You must additionally have admin permissions to the node to connect an OSFGroup to a node.
- group members belong to an OSFGroups member group.  group managers belong to both the manager and member group.  We organize this way to help us achieve giving all OSFGroup members the same permission to a node.  Internally, the OSFGroup's "member" Django group is given perms to a node. So all users in the member group (which are members + managers) will inherit the permissions of that django group, and all have write perms to a node, for example.


## Important Notes
1.  "Contributors" are traditional project contributors, what we've always had.  On the other hand, if you have your perms through group membership, you're a "group member".  You can also be both, and this scenario will be common (an admin contributor adds their OSF Group to the node, for example). When adding OSF groups, the goal was to make OSF Groups members as similar to contributors as possible.  Nearly everything a contributor can do, a group member can do.  Exceptions are creating registrations of projects, and the fact that we must always have one registered admin *contributor* (not group member) per project.  Also, group members don't show up in the project citation.
2. Generally, checking a user's permissions or if a user has a certain permission, will return your highest permission, either through group membership or contributorship. Where you care that a user is a contributor specifically, use Node.is_contributor() or Node.is_admin_contributor().
3.  With OSF Groups, there are many more ways a user can gain/lose permissions to a project.  Aside from being removed as a contributor, if they have an OSF Group connected to a project, their 1) group can be removed from the project 2) they can be removed as a member of the group or 3) their group can be deleted altogether.  The same kinds of checks we do when removing contributors (ex. file checkin, unsubscribing from project subscriptions, disconnecting addons) may need to be performed, provided the user has lost permissions through every avenue.
4. Use guardian shortcut `get_group_perms` for determining which perms the user has to a node. This infers permissions through the user's group membership and ignore superuser permissions by default. This will return both perms you have because you are a traditional contributor or a group member (you belong to a Node's read/write/admin django group, or you belong to an OSFGroup's Django member group that has permissions to the node)
5. Use caution in the areas where contributor versus group member permissions matter.  For example, if I am a write contributor, but an admin group member, I do **not** satisfy the one admin contributor per node rule, even though I am a contributor and I have admin perms.  Make sure the user has admin perms through contributorship specifically. (Node.is_admin_contributor(user) check will take care of this for you.)
6. Don't confuse "OSFGroup" language with "Django groups".   This PR introduces Groups (OSF Groups) functionality, allowing a user to add a collection of users, like their lab, to a project all at once.  We use Django groups to make that possible internally.  Externally, to our users, we will just call OSFGroups, "Groups", and won't mention Django groups at all.
7.  Generally, you give permissions to django groups, and add users to these groups.  To see if a user has permission, though, check the user. However, there are definitely times where we have to check Django group membership.  For example, places where we want to know an unregistered contributor's intended permission (they have no permissions until they claim their account, but we can check their Django group membership to see they will have write perms to the node when they've registered).  Another useful place where we check Django group membership instead of user permissions, is where we care whether the user got permissions from belonging to an OSFGroup that had permissions to the node, or they were a contributor on the node.
8. Tricky areas: merging users with both group member perms/contributor perms, gdpr delete, inheriting contributors/group members when creating a component.  Claiming accounts - when account is claimed, info is updated across all nodes and osf groups.  Marking users as spam, all nodes user is connected to (contributor/group member perms) are marked private

# Post Guardian-Groups Permissions Guide:
### Node Model
- *Node.objects.can_view(...)* Returns all the nodes that the user can view.  This means public nodes, non-deleted nodes, as well as any nodes you have read permissions on, or nodes whose parents you have admin permissions on.  You can have these permissions by being a traditional contributor OR a group member.  Does not include nodes whose perms you only have because you're a superuser.
- _Node.get_permissions(self, user)_  Returns a list or explicit permissions, lowest to highest, of the perms the user has to the node (contributor and group member), superusers excluded. Ex. ['read', 'write'].  Implicit admin perms not included here.
- _Node.has_permission(self, user, permission, check_parent=True)_ Determines if the user has the specified permission to the node, either by being a contributor or group member.  If check_parent is True, checks to see if you have admin permissions on the parent, either through contributorship, or group membership.   Superuser perms are not factored in here.
- _Node.is_contributor(user)_ Returns true if the user is a traditional contributor **only**. Superusers excluded, implicit admin perms excluded.
- _Node.is_contributor_or_group_member(user)_ Returns true if the user is a traditional contributor or is a group member on the node. Superuser perms not factored in, implicit admin not included here.
- _Node.is_admin_contributor(user)_ - Returns true if the user has admin perms on the node, and they are an admin *contributor*.  Admin group membership doesn't count.  There are certain things that only admin contributors can do, like create a registration, and we need at least one registered admin contrib on the node at all times.
- Node.can_view() - Returns True if the user can view the node. Calls `.has_permission`, so same rules there apply, group members included.
- _Node.contributors_and_group_members_ property - Returns a queryset of users who are contributors and/or group members to the node - no duplicates. Implicit admin not factored in, no superusers.

### User Model 
- _User.get_projects_in_common(other_user)_ now returns projects that both users are either contributors or group members on 
- _User.contributor_to_ - Returns nodes that the user only has perms on due to their contributorship, not group membership
- _User.contributor_or_group_member_to_ -  Returns nodes the user has perms on, either through group membership or contributorship

### APIv2
- _NodeSerializer.current_user_permissions_ Returns a list, lowest to highest, with the permissions logged-in user has to the given node.  These are either contributor permissions or group member permissions, superuser perms excluded.  Highest perm displayed. Implicit admin factored in; if you are an implicit admin contrib or group member, you have read perms to the given node.
- _NodeContributor endpoints_ - contributor permissions displayed are the perms the user has because of their contributorship, **NOT** group membership.  The distinction between contributor permissions and group membership permissions are important where it comes to managing contributors on a node. For example, a group member might have write perms on a node, but not be a contributor.  We don't want to rely on their write perms to determine if they are already a contributor.  Another example is that a group member might have admin perms, but be a read contributor, so this user is not sufficient to make sure the one admin *contributor* per node rule is enforced.
- _OSFGroup endpoints_ - `v2/groups/`, `v2/groups/<group_id>`, `vs/groups/members` - serializes OSF Groups, along with their members.  Users can have member or manager roles in the group.  Use these endpoints to manage membership of the OSFGroups themselves.
- _NodeGroup endpoints_ - `v2/nodes/node_id/groups`, `v2/nodes/node_id/groups/group_id` - serializes OSF groups that are attached to the node.  Displays the permission that OSF group has to the node.  Every member of the OSF Group is given the same permission to the node.  Use these endpoints to update which groups are attached to the node, and display/update the group's permissions.
- _UserGroup endpoints_ - Displays the OSFGroups the user belongs to
- _UserNodes_ endpoint - Displays all nodes the user belongs to, either through contributorship or group membership, so all will show up on your dashboard.
- _NodeContributorsAndGroupMembersList_ - returns users who are contributors and/or group members on the node.

# Handy OSFGroup-related properties and methods - 
- _Node.osf_groups_ property - returns a queryset of OSFGroups that have some permission to the node
- _Node.get_osf_groups_with_perms(self, permission) method_ - returns a queryset of all OSFGroups that have the specified permission to the node - implicit admin not factored in
- _OSFGroup.nodes property_ - returns a queryset of all nodes that the OSFGroup has permissions to 
- _OSFGroup.get_permission_to_node(node) method_ - If the OSFGroup is connected to the node, returns the permission the OSFGroup has - implicit admin not factored in
- _User.osf_groups property_ - Returns a queryset of the OSFGroups that a user belongs to
- _User.group_role(group) method_ - Returns the role that a user has in an OSFGroup


## QA Notes

The main thing that this PR does is fundamentally change how permissions are managed, so we need to be thorough about this. Regarding the OSFGroups section specifically, this is just backend for groups, so little to test there. There were lots of API endpoints added for OSF Groups, so this can be added to Runscope, if you want, see documentation link below.  Or you can wait to add runscope stuff when frontend goes in, as groups API might change slightly.

### Data migration 
-  Make a bunch of projects/components and give many contributors with different permissions as well as unregistered contributors.  Make sure the contributors are all there post-migration, with the same perms.  Do the same for some preprint contributors.  Also make a few registrations. 

### OSF 

I would do a very thorough regression of the entire OSF (sorry!), as well as a light regression of Preprints.  Pay special attention to permissions - behavior should be the same as before.
 In particular:
- creating projects
- creating components (and copying contributors from parent)
- merging users
- managing contributors on a project
- adding unregistered contributors
- comments on projects - especially at mentions
- claiming contributors
- viewing contributors that are admins on a parent project
- searching for contributors to add to a project, # projects in common
- viewing/updating node settings
- managing files
- connecting addons
- Removing contribs - should check files back in, remove email subscriptions, disconnect addons in most cases
- Editing wikis
- Creating registrations
- Creating forks/templating
- Making projects public/private
- Editing project title, description, tags, license?



## Documentation

https://github.com/CenterForOpenScience/developer.osf.io/pull/33
## Side Effects
- This basically gives a user many more ways in which they can have permissions to a node.  I can be a write contributor, belong to Group A which has read permissions on the node, and Group B which has write permissions on the node.  We usually care about the highest permission, but sometimes their traditional contributor permission is important.  Also, we have to be careful when removing a contributor from a node, that we don't, for example, check all their files in, if they still have group membership.

- My main concern is making things more slooooow.  Some actions are time-consuming, like removing an OSFGroup from a node, because there are a lot of things happening there.

- Over 40% of the files changed in this PR are tests. We were really thorough with test coverage, for such an important change.

## Ticket
[Guardian Epic](https://openscience.atlassian.net/browse/PLAT-1041
)
[Groups Epic](https://openscience.atlassian.net/browse/PLAT-1185
)
[Final Polishing Ticket](https://openscience.atlassian.net/browse/PLAT-1268)

pattisdr commented on website/search/search.py:
```diff
@@ -61,6 +61,23 @@ def update_preprint(preprint, index=None, bulk=False, async_update=True, saved_f
         return search_engine.update_preprint(preprint, **kwargs)
 
+@requires_search
+def update_group(group, index=None, bulk=False, async_update=True, saved_fields=None, deleted_id=None):
```
Passing in deleted_id here, because currently when an OSFGroup is deleted, it's literally deleted, not marked as deleted.

pattisdr commented on osf/models/osf_group.py:
```diff
@@ -0,0 +1,557 @@
+    """
+
+    name = models.TextField(blank=False)
+    creator = models.ForeignKey(OSFUser,
```
Similar to creator on the Node model, this field has null=True, so if the creator is deleted, this can be set to null.  However, an OSFGroup needs a creator.

pattisdr commented on addons/osfstorage/listeners.py:
```diff
@@ -11,9 +11,25 @@
+    """ Listens to a contributor or group member being removed to check in all of their files
     """
-    node.files.filter(checkout=user).update(checkout=None)
+    enqueue_postcommit_task(checkin_files_task, (node._id, user._id, ), {}, celery=True)
```
Checking in files turned into a celery task, similar to deleting files.

pattisdr commented on addons/base/views.py:
```diff
@@ -331,7 +331,7 @@ def get_auth(auth, **kwargs):
     if waffle.switch_is_active(features.ELASTICSEARCH_METRICS):
         user = auth.user
-        if isinstance(node, Preprint) and not node.is_contributor(user):
+        if isinstance(node, Preprint) and not node.is_contributor_or_group_member(user):
```
Even though preprints don't have OSF Groups at the moment, they may one day, and this allows that possibility.  For now, it will work for contributors only.

pattisdr commented on admin/users/serializers.py:
```diff
@@ -3,12 +3,19 @@
         'name': user.fullname,
         'id': user._id,
-        'nodes': list(map(serialize_simple_node, user.contributor_to)),
+        'nodes': list(map(serialize_simple_node, user.contributor_or_group_member_to)),
```
Now listing all nodes in admin app that user is a contributor or group member to, not just contributor to.

pattisdr commented on api/nodes/views.py:
```diff
@@ -570,19 +580,14 @@ class NodeDraftRegistrationsList(JSONAPIBaseView, generics.ListCreateAPIView, No
-            branched_from=node,
-            deleted__isnull=True,
-        )
+        return node.draft_registrations_active
```
This property already defined on node.

Johnetordoff commented on addons/base/utils.py:
```diff
@@ -31,7 +53,7 @@ def get_addons_by_config_type(config_type, user):
     msg = """
     </div>"""  # None is default
-    if error_type != 'FILE_SUSPENDED' and ((auth.user and node.is_contributor(auth.user)) or (auth.private_key and auth.private_key in node.private_link_keys_active)):
+    if error_type != 'FILE_SUSPENDED' and ((auth.user and node.is_contributor_or_group_member(auth.user)) or (auth.private_key and auth.private_key in node.private_link_keys_active)):
```
slight nit: this conditional is super long and would be more readable if split over multiple lines.

Johnetordoff commented on api/osf_groups/serializers.py:
```diff
@@ -0,0 +1,191 @@
+                else:
+                    user = group.add_unregistered_member(full_name, email, auth, role)
+        except ValueError as e:
+            # Making someone a member has the potential to violate the one-manager rule
```
This comment isn't clear to me, it seems like making you mean taking away a manger role and making that a member role, but aren't all managers also members? Also I'm unsure where the ValueError is being thrown. I'd separate the registered and unregistered user code into two separate clauses, even if it's slightly longer.

Johnetordoff replied:
Sanity check: adding an unregistered member can't possibly violate the one manager rule, because you need one manager to even have a group to begin with.

pattisdr replied:
Yes, thank you, this comment doesn't describe why this is needed here.  (Background info: managers are also members, but they have the 'manager' permission while members have the 'member' permission.  If I'm a manager and I run OSFGroup.make_member(...) on myself, I am downgrading my manager perms. But that's not possible here).

However, both adding a member or an unregistered member can raise ValueErrors and I wanted to share this with both.   For example, if I'm trying to add an unregistered member that already exists, will raise a ValueError on the OSFGroup model.

Johnetordoff commented on api/osf_groups/views.py:
```diff
@@ -0,0 +1,231 @@
+        serializer.save(creator=user)
+
+
+class GroupDetail(JSONAPIBaseView, generics.RetrieveUpdateDestroyAPIView, OSFGroupMixin):
```
Again you might want to consolidate  these (`GroupDetail` and `GroupList`) to inherit from a base class, the static elements are very similar and future additions probably won't deviate from them much.

pattisdr replied:
ok, I try to weigh DRY versus readable, and if only a little code is consolidated, I don't find it's always worth it.  I can consolidate this, however.

Johnetordoff commented on api/osf_groups/views.py:
```diff
@@ -0,0 +1,231 @@
+            raise NotFound('{} cannot be found in this OSFGroup'.format(user._id))
+
+    def get_serializer_class(self):
+        if self.request.method == 'PUT' or self.request.method == 'PATCH' or self.request.method == 'DELETE':
```
might just want to make `method = self.request.method` to save some space or `self.request.method  in ('PUT', 'PATCH', 'DELETE')`

Johnetordoff commented on api/osf_groups/serializers.py:
```diff
@@ -0,0 +1,191 @@
+        auth = Auth(self.context['request'].user)
+
+        try:
+            # Making sure the one-manager rule isn't violated
```
I understand in this context you don't want to down grade a manager by making them a member, thus removing all managers. 👍

Johnetordoff commented on api/osf_groups/views.py:
```diff
@@ -0,0 +1,231 @@
+        return resource_object_list
+
+    # Overrides ListFilterMixin
+    def build_query_from_field(self, field_name, operation):
```
I feel something like this is a bit cleaner 
```python
    def build_query_from_field(self, field_name, operation):
        
        if field_name != 'role':
            return super(GroupMembersList, self).build_query_from_field(field_name, operation)
        if operation['op'] != 'eq':
            raise InvalidFilterOperator(value=operation['op'], valid_operators=['eq'])
        
        role = operation['value'].lower().strip()
        if role not in GROUP_ROLES:
            raise InvalidFilterValue(value=operation['value'])
        else:
            group = self.get_osf_group(check_object_permissions=False)
            return Q(id__in=group.managers if role == MANAGER else group.members_only)
```

pattisdr replied:
Thanks, I borrowed some pieces from this.  I prefer the first thing to not be a negative, if A: do this, otherwise, do this.  Instead of if not A: move on.

pattisdr commented:
Excellent @Johnetordoff ! I certainly appreciate you taking time to go through this today - will definitely be more polished after I make these changes.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -50,6 +51,24 @@ def has_object_permission(self, request, view, obj):
         return obj.is_public or obj.can_view(auth)
 
 
+class IsAdminContributor(permissions.BasePermission):
```
This is from product - Because membership of a group can change, we are not copying the group over when a node is registered.  So because that group member is not going to be on the node, we're not allowing them to register or create draft registrations.  But there's no reason to prevent them from being viewing.

A group admin can't edit a draft, but if they're made a regular admin contrib, they can.  You can also be both a group member and a contributor, in fact, that will be common. You don't need to be removed from the group and readded as a contrib.

pattisdr commented on addons/osfstorage/tests/test_models.py:
```diff
@@ -841,7 +841,7 @@ def setUp(self):
     def test_checkout_logs(self):
         non_admin = factories.AuthUserFactory()
-        self.node.add_contributor(non_admin, permissions=['read', 'write'])
+        self.node.add_contributor(non_admin, permissions='write')
```
yeah, I didn't bother importing the constants in places like this, but I'll try to change the majority of them!

pattisdr commented on api/nodes/filters.py:
```diff
@@ -78,17 +80,24 @@ def build_query_from_field(self, field_name, operation):
+            if user.is_anonymous:
+                # Anonymous users have no perms to the current node in current versions, and in
+                # older versions, will have read if node is public
+                query = Q() if StrictVersion(self.request.version) < StrictVersion('2.11') and perm == READ else Q(id__in=[])
```
Sure, I can avoid defining query.

pattisdr commented on api/nodes/filters.py:
```diff
@@ -78,17 +80,24 @@ def build_query_from_field(self, field_name, operation):
             return query
         return super(UserNodesFilterMixin, self).build_query_from_field(field_name, operation)
+
+    def build_node_list(self, user, perm):
```
Sure, I can't think of when we'd need with_superuser to be True here, but it doesn't hurt to make it a variable.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -50,6 +51,24 @@ def has_object_permission(self, request, view, obj):
+    admin contributor to make changes.  Admin group membership
+    is not sufficient.
+    """
+    acceptable_models = (AbstractNode, DraftRegistration)
```
Yes, product choice, one of the few distinctions b/t group mems and contribs.  Admin group membership being insufficient, doesn't have anything to do with the acceptable_models variable.  `acceptable_models` is just a tuple for what objects this Permission class is meant to be used for.  It's not off-limits for modification;  if someone wants to expand this for other models, they can.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -50,6 +51,24 @@ def has_object_permission(self, request, view, obj):
+
+    def has_object_permission(self, request, view, obj):
+        assert_resource_type(obj, self.acceptable_models)
+        if isinstance(obj, DraftRegistration):
```
Because it is possible acceptable_models are expanded, I don't want to assume it's either/or.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -112,6 +129,21 @@ def has_object_permission(self, request, view, obj):
 
+class AdminContributorOrPublic(permissions.BasePermission):
+
+    acceptable_models = (AbstractNode, DraftRegistration)
```
No need to be a frozenset, they can expand this tuple if they want!

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -112,6 +129,21 @@ def has_object_permission(self, request, view, obj):
+        """
+        To make changes, user must be an admin contributor. Admin group membership is not sufficient.
+        """
+        assert_resource_type(obj, self.acceptable_models)
```
This is for the developer, not the user, and I think it's important we fail noticeably so the dev can modify the permission class or make sure their view is returning the right item/checking permissions in the right place if they didn't expect it to fail.

When it fails, you can add your model to acceptable models, but you might have to modify the method to handle your new model as well.   

(An example is the DraftRegistration object. DraftRegistrations don't have contributors attached, but they're `branched_from`, (node) field does.  So a mod was added to pull the branched_from field if you needed it.)

This is a pattern that we've used for awhile, and I've found this check to be really helpful in development.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -112,6 +129,21 @@ def has_object_permission(self, request, view, obj):
+        assert_resource_type(obj, self.acceptable_models)
+        auth = get_user_auth(request)
+        if request.method in permissions.SAFE_METHODS:
+            return obj.is_public or obj.can_view(auth)
```
It's not entirely redundant, as you allude to in your other comment, because looking through Node.can_view(...), if I'm viewing the node through an anonymous, private link but that link isn't attached to this node, then this would return False.  But we want the is_public check to bypass everything.  The Node.can_view() method may change in the future, but if it's public, regardless of any other conditionals in can_view, we want them to be able to view this.

Regarding the use of ternary statements, my personal preference is to avoid them. It's less lines, but sometimes readability is sacrificed, and I like being able to quickly put a breakpoint in there if need be.   If it's a very succinct ternary, I may use it.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -148,6 +180,29 @@ def has_object_permission(self, request, view, obj):
+    """Permissions for node group detail - involving who can update the relationship
+    between a node and an OSF Group."""
+
+    acceptable_models = (OSFGroup, AbstractNode,)
```
This reminded me that I wanted trailing commas for all of them.  Trailing commas are optional in multi-element tuples: I prefer them for when you start breaking this tuple out in multiple lines, plus ease in adding new members to tuples.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -148,6 +180,29 @@ def has_object_permission(self, request, view, obj):
+        auth = get_user_auth(request)
+        node = self.load_resource(request.parser_context['kwargs'], view)
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
```
I don't think we should rely on can_view always remaining the same.  I think asking 1) is it public? or 2) do you have specific permission to view is more readable and preserves the main intent of this permission class.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -148,6 +180,29 @@ def has_object_permission(self, request, view, obj):
+        node = self.load_resource(request.parser_context['kwargs'], view)
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        elif request.method == 'DELETE':
```
Yes, we're removing the group from the node, not literally deleting the group.  I'm broadly thinking of this as a delete operation, we're removing permissions to the Node from the OSF Group's Django member group.  Plus, because DELETE and PATCH operations need different permissions, separating this way makes it easy to do this from a high-level. Finally, this is consistent with many of our other similar operations, deleting a contributor from a node for example. We're not technically deleting the user object, but we're removing the relationship between that user and the node, plus the user's permission to that node, and whether the user is bibliographic.

pattisdr commented on api/nodes/permissions.py:
```diff
@@ -134,6 +166,29 @@ def has_object_permission(self, request, view, obj):
+            # or you need to be an OSF group manager
+            return node.has_permission(auth.user, osf_permissions.ADMIN) or obj.has_permission(auth.user, 'manage')
+        else:
+            return node.has_permission(auth.user, osf_permissions.ADMIN)
```
idk, I value consistency a lot in these patterns we use often.  We're often doing something like this, 1)is it a safe method 2) Is it a delete method? 3) Otherwise do this for all other methods.  Every element of the conditional is related to the method.  

I personally find this less readable: 1) are you an admin? 2) is this a delete method ...and so on...they are different questions.

pattisdr commented on api/users/views.py:
```diff
@@ -310,11 +312,20 @@ class UserNodes(JSONAPIBaseView, generics.ListAPIView, UserMixin, UserNodesFilte
     def get_default_queryset(self):
         user = self.get_user()
+        # Nodes the requested user has read_permissions on
+        default_queryset = user.contributor_or_group_member_to.filter(type__in=['osf.node'])
```
yep!

pattisdr commented on api/users/views.py:
```diff
@@ -325,6 +335,32 @@ def get_queryset(self):
+        current_user = self.request.user
+        if current_user.is_anonymous:
+            return OSFGroup.objects.none()
+        return requested_user.osf_groups.filter(id__in=current_user.osf_groups.values_list('id', flat=True))
```
You can access a group directly that you don't belong to, for example, if you see a group attached to a project, you can access that group. I was originally showing all groups under `/v2/groups/`, but was told to restrict it.   But yes, we may revisit this in the future.

pattisdr commented on api/users/serializers.py:
```diff
@@ -142,7 +147,7 @@ class Meta:
         user = get_user_auth(self.context['request']).user
         if obj == user:
-            return user.contributor_to.count()
+            return user.contributor_or_group_member_to.filter(type__in=['osf.node']).count()
```
yep! I was copying and pasting from somewhere else and missed this. thanks.

pattisdr commented on api/osf_groups/permissions.py:
```diff
@@ -0,0 +1,39 @@
+    def has_object_permission(self, request, view, obj):
+        if not isinstance(obj, OSFGroup):
+            obj = OSFGroup.load(request.parser_context['kwargs']['group_id'])
+        assert_resource_type(obj, self.acceptable_models)
```
I think improvements to `assert_resource_type` specifically are beyond the scope of this PR.  Feel free to reticket - this is a pattern we use on most of our permission classes, so I don't want to mix this in with this giant PR.

pattisdr commented on api/nodes/views.py:
```diff
@@ -1156,6 +1161,104 @@ def get_object(self):
+        return context
+
+
+class NodeGroupsDetail(JSONAPIBaseView, generics.RetrieveUpdateDestroyAPIView, NodeMixin, OSFGroupMixin):
```
Sure, but besides the base classes, a List and a Detail endpoint aren't going to share a lot of the same methods.  I DRY them up when it's a significant amount of shared code, like the `OSFGroupMemberBaseView` in this same PR.  But it doesn't hurt to dry them up here, I just like to be careful about the tradeoffs between being DRY and being readable.

Johnetordoff replied:
I think it's nice to share a base class in a lot of these because it makes it really obvious how the sub classes static members and inherited classes are different hence readability as a opposed to just DRY-ness, but these could go either way.

pattisdr commented on api/osf_groups/serializers.py:
```diff
@@ -0,0 +1,191 @@
+
+    def update(self, instance, validated_data):
+        if 'name' in validated_data:
+            instance.set_group_name(validated_data.get('name'))
```
You'll get a "This field can't be None, or similar" `name` is a CharField with required=True, so DRF will enforce that it can't be None or an empty string. That happens before this update method is called.

pattisdr commented:
Addressed your comments @Johnetordoff - Main change was the ADMIN/WRITE/READ constant usage - it was quicker to make this change by replacing the majority of the strings across the entire codebase, so this added a lot of files to this PR.

pattisdr commented:
Getting this up-to-date - recent merge to develop added queries that need to be updated for guardian.

pattisdr commented on addons/base/utils.py:
```diff
@@ -1,9 +1,31 @@
+                    'category': markupsafe.escape(node.category_display),
+                    'title': markupsafe.escape(node.title),
+                    'user': markupsafe.escape(user.fullname)
+                })
```
I see what you're saying - I'd moved it to here to call on various OSFGroup methods, but I need the node anyway! So yes, makes sense to move to Node model.

pattisdr commented:
Recent commits speed up node serialization by annotating preprint and region info

pattisdr commented on website/static/js/myProjects.js:
```diff
@@ -1965,7 +1965,7 @@ var Information = {
             var item = args.selected()[0].data;
             var resourceType = item.type;
-            var permission = item.attributes.current_user_permissions.slice(-1)[0];
+            var permission = item.attributes.current_user_permissions[0];
```
Not relying on list order is https://github.com/CenterForOpenScience/osf.io/pull/8957/files

pattisdr commented on osf/models/node.py:
```diff
@@ -344,17 +368,6 @@ def parent_node(self):
-            return [] if self.annotated_tags == [None] else self.annotated_tags
-        else:
-            return self.tags.values_list('name', flat=True)
-
```
Steve and I looked into this - this tag_names optimization was actually slowing things down after groups/guardian - we didn't know if it was an interaction with guardian, but this "optimization" has been removed.

mfraezz commented on osf/migrations/0164_add_guardian_to_nodes.py:
```diff
@@ -0,0 +1,200 @@
+# Forward migration - for each node, create a read, write, and admin Django group
+add_node_read_write_admin_auth_groups = """
+    INSERT INTO auth_group (name)
+    (SELECT 'node_' || N.id || '_read' FROM osf_abstractnode AS N WHERE N.id > {start} AND N.id <= {end}
```
Minor: This subquery can be done in one query with something like
```
SELECT regexp_split_to_table('node_' || N.id || '_read,node_' || N.id || '_write,node_' || N.id || '_admin', ',')
FROM osf_abstractnode N
WHERE N.id > {start} AND N.id < {end};
```
Benchmarked with pages of 10k, planning averaged ~70% faster, execution averaged ~3% faster (total of ~5% faster)

pattisdr replied:
Total migration taking about 20 min w/ this change - this time the commit phase was maybe 6-7 minutes, but the other phase shortened by ~2 min.

mfraezz commented on osf/migrations/0164_add_guardian_to_nodes.py:
```diff
@@ -0,0 +1,200 @@
+    AND N.id > {start}
+    AND N.id <= {end};
+
+    -- Adds "read_node" and "write_node" permissions to all Node write groups
```
Can these two queries be consolidated with: 
```
AND (
  PERM.codename = 'read_node' OR
  PERM.codename = 'write_node'
) [...]
``` 
?
(Disclaimer: I have not checked to see if this works and is faster)

pattisdr replied:
Made this change, seems to work, no major changes noticed in speed.

pattisdr commented on addons/osfstorage/listeners.py:
```diff
@@ -11,9 +11,27 @@
+
+
+@app.task(max_retries=5, default_retry_delay=60)
+def checkin_files_task(node_id, user_id):
```
I don't think that's needed - when the `contributor_removed` signal is fired, the `checkin_files_task` method runs, which then queues this `checkin_files_task`.  I've verified locally, but let me know if there's something I'm missing!

pattisdr commented:
Temporary migration script added to verify counts scripts/remove_after_use/verify_groups_guardian_migration.py

pattisdr commented on api/nodes/serializers.py:
```diff
@@ -480,37 +488,44 @@ class NodeSerializer(TaxonomizableSerializerMixin, JSONAPISerializer):
-            if not permissions and user in obj.parent_admin_contributors:
-                permissions += ['read']
-            return permissions
+            user_perms = obj.get_permissions(user)[::-1]
```
Note: annotations for optimizations removed for most purposes - was slowing down node lists quite a bit.
